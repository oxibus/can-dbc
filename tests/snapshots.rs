#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
