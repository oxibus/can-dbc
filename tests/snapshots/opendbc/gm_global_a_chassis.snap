---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - K182_PACM
  - K43_PSCM
  - K17_EBCM
  - NEO
  - K124_ASCM
messages:
  - id:
      Standard: 823
    name: PACMParkAssitCmd
    size: 7
    transmitter:
      NodeName: NEO
    signals:
      - name: RollingCounter
        multiplexer_indicator: Plain
        start_bit: 35
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: SteeringWheelChecksum
        multiplexer_indicator: Plain
        start_bit: 47
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: SteeringWheelCmd
        multiplexer_indicator: Plain
        start_bit: 23
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
  - id:
      Standard: 560
    name: EBCMRegen
    size: 6
    transmitter:
      NodeName: K17_EBCM
    signals:
      - name: Regen
        multiplexer_indicator: Plain
        start_bit: 1
        size: 10
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
  - id:
      Standard: 368
    name: EBCMFrictionBrakeStatus
    size: 8
    transmitter:
      NodeName: K17_EBCM
    signals:
      - name: FrictionBrakePressure
        multiplexer_indicator: Plain
        start_bit: 23
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
  - id:
      Standard: 789
    name: EBCMFrictionBrakeCmd
    size: 5
    transmitter:
      NodeName: K17_EBCM
    signals:
      - name: RollingCounter
        multiplexer_indicator: Plain
        start_bit: 33
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: FrictionBrakeMode
        multiplexer_indicator: Plain
        start_bit: 7
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: FrictionBrakeChecksum
        multiplexer_indicator: Plain
        start_bit: 23
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
      - name: FrictionBrakeCmd
        multiplexer_indicator: Plain
        start_bit: 3
        size: 12
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - NEO
message_transmitters:
  - message_id:
      Standard: 823
    transmitter:
      - NodeName: K43_PSCM
      - NodeName: NEO
  - message_id:
      Standard: 789
    transmitter:
      - NodeName: NEO
      - NodeName: K17_EBCM
comments:
  - Node:
      name: K182_PACM
      comment: Parking Assist Control Module
  - Node:
      name: K43_PSCM
      comment: Power Steering Control Module
  - Node:
      name: K17_EBCM
      comment: Electronic Brake Control Module
  - Node:
      name: NEO
      comment: Comma NEO
  - Node:
      name: K124_ASCM
      comment: Active Safety Control Module
attribute_definitions:
  - Plain: "\"UseGMParameterIDs\" INT 0 0"
  - Plain: "\"ProtocolType\" STRING"
  - Plain: "\"BusType\" STRING"
attribute_defaults:
  - name: UseGMParameterIDs
    value:
      Double: 1
  - name: ProtocolType
    value:
      String: GMLAN
  - name: BusType
    value:
      String: ""
attribute_values:
  - name: UseGMParameterIDs
    value:
      Raw:
        Double: 0
  - name: BusType
    value:
      Raw:
        String: CAN
  - name: ProtocolType
    value:
      Raw:
        String: GMLAN
